quinn-proto/src/connection/cid_state.rs

```diff
@@ -0,0 +1,207 @@
+//! Maintain the state of local connection IDs
+use std::{
+    collections::{HashSet, VecDeque},
+    time::{Duration, Instant},
+};
+
+use tracing::{debug, trace};
+
+use crate::{shared::IssuedCid, TransportError};
+
+/// Local connection ID management
+pub struct CidState {
+    /// Timestamp when issued cids should be retired
+    retire_timestamp: VecDeque<CidTimestamp>,
+    /// Number of local connection IDs that have been issued in NEW_CONNECTION_ID frames.
+    issued: u64,
+    /// Sequence numbers of local connection IDs not yet retired by the peer
+    active_seq: HashSet<u64>,
+    /// Sequence number the peer has already retired all CIDs below at our request via `retire_prior_to`
+    prev_retire_seq: u64,
+    /// Sequence number to set in retire_prior_to field in NEW_CONNECTION_ID frame
+    retire_seq: u64,
+    /// cid length used to decode short packet
+    cid_len: usize,
+    //// cid lifetime
+    cid_lifetime: Option<Duration>,
+}
+
+impl CidState {
+    pub(crate) fn new(cid_len: usize, cid_lifetime: Option<Duration>, now: Instant) -> Self {
+        let mut active_seq = HashSet::new();
+        // Add sequence number of CID used in handshaking into tracking set
+        active_seq.insert(0);
+        let mut this = CidState {
+            retire_timestamp: VecDeque::new(),
+            issued: 1, // One CID is already supplied during handshaking
+            active_seq,
+            prev_retire_seq: 0,
+            retire_seq: 0,
+            cid_len,
+            cid_lifetime,
+        };
+        // Track lifetime of cid used in handshaking
+        this.track_lifetime(0, now);
+        this
+    }
+
+    /// Find the next timestamp when previously issued CID should be retired
+    pub(crate) fn next_timeout(&mut self) -> Option<Instant> {
+        self.retire_timestamp.front().map(|nc| {
+            trace!("CID {} will expire at {:?}", nc.sequence, nc.timestamp);
+            nc.timestamp
+        })
+    }
+
+    /// Track the lifetime of issued cids in `retire_timestamp`
+    pub(crate) fn track_lifetime(&mut self, new_cid_seq: u64, now: Instant) {
```

Comment:
I think this is purely internal now:
```suggestion
    fn track_lifetime(&mut self, new_cid_seq: u64, now: Instant) {
```